azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -1870,57 +1870,58 @@ private void selectExecutorAndDispatchFlow(final ExecutionReference reference,
         throws ExecutorManagerException {
       final Set<Executor> remainingExecutors = new HashSet<>(ExecutorManager.this.activeExecutors);
       synchronized (exflow) {
-        for (int i = 0; i <= this.maxDispatchingErrors; i++) {
-          final String giveUpReason = checkGiveUpDispatching(reference, remainingExecutors);
-          if (giveUpReason != null) {
-            logger.error("Failed to dispatch queued execution " + exflow.getId() + " because "
-                + giveUpReason);
-            finalizeFlows(exflow);
-            // GIVE UP DISPATCHING - exit
+        do {
+          final Executor selectedExecutor = selectExecutor(exflow, remainingExecutors);
+          if (selectedExecutor == null) {
+            ExecutorManager.this.commonMetrics.markDispatchFail();
+            handleNoExecutorSelectedCase(reference, exflow);
+            // RE-QUEUED - exit
             return;
           } else {
-            final Executor selectedExecutor = selectExecutor(exflow, remainingExecutors);
-            if (selectedExecutor == null) {
-              ExecutorManager.this.commonMetrics.markDispatchFail();
-              handleNoExecutorSelectedCase(reference, exflow);
-              // RE-QUEUED - exit
+            try {
+              dispatch(reference, exflow, selectedExecutor);
+              ExecutorManager.this.commonMetrics.markDispatchSuccess();
+              // SUCCESS - exit
               return;
-            } else {
-              try {
-                dispatch(reference, exflow, selectedExecutor);
-                ExecutorManager.this.commonMetrics.markDispatchSuccess();
-                // SUCCESS - exit
-                return;
-              } catch (final ExecutorManagerException e) {
-                logFailedDispatchAttempt(reference, exflow, selectedExecutor, e);
-                ExecutorManager.this.commonMetrics.markDispatchFail();
-                reference.setNumErrors(reference.getNumErrors() + 1);
-                // FAILED ATTEMPT - try other executors except selectedExecutor
-                remainingExecutors.remove(selectedExecutor);
-              }
+            } catch (final ExecutorManagerException e) {
+              logFailedDispatchAttempt(reference, exflow, selectedExecutor, e);
+              ExecutorManager.this.commonMetrics.markDispatchFail();
```

Comment:
should this case count as dispatch failure?

Reply:
Sorry, what so you mean?

Reply:
i mean since the flow will still be dispatched in this case, still marking this as dispatch fail is a bit confusing. IMO we should distinguish retryable & non-retryable failure, and only mark dispatch failure for non-retryable cases.

Reply:
Someone might set up alert on dispatching failure, like us. So we want to make sure alert indicates real dispatching failures.

Reply:
Yes. But as I said, this matches the existing behaviour, I'm not changing it. For example my proposal of splitting to different counters could be done later if desired.